emle/models.py

```diff
@@ -1210,3 +1223,394 @@ def forward(self, atomic_numbers, charges_mm, xyz_qm, xyz_mm):
+            shifts = _torch.mm(shifts_idx, cell)
+        else:
+            shifts = _torch.zeros(
+                (edge_index.shape[1], 3), dtype=self._dtype, device=positions.device
+            )
+
+        return edge_index, shifts
+
+    def to(self, *args, **kwargs):
+        """
+        Performs Tensor dtype and/or device conversion on the model.
+        """
+        module = super(MACExEMLE, self).to(*args, **kwargs)
+        module._mace = module._mace.to(*args, **kwargs)
+        return module
+
+    def cpu(self, **kwargs):
+        """
+        Returns a copy of this model in CPU memory.
+        """
+        module = super(MACExEMLE, self).cpu(**kwargs)
+        module._mace = module._mace.cpu(**kwargs)
+        if self._atomic_numbers is not None:
+            module._atomic_numbers = module._atomic_numbers.cpu(**kwargs)
+        return module
+
+    def cuda(self, **kwargs):
+        """
+        Returns a copy of this model in CUDA memory.
+        """
+        module = super(MACExEMLE, self).cuda(**kwargs)
+        module._mace = module._mace.cuda(**kwargs)
+        if self._atomic_numbers is not None:
+            module._atomic_numbers = module._atomic_numbers.cuda(**kwargs)
+        return module
+
+    def double(self):
+        """
+        Returns a copy of this model in float64 precision.
+        """
+        module = super(MACExEMLE, self).double()
+        module._mace = module._mace.double()
+        return module
+
+    def float(self):
+        """
+        Returns a copy of this model in float32 precision.
+        """
+        module = super(MACExEMLE, self).float()
+        module._mace = module._mace.float()
+
+        return module
+
+    def forward(self, atomic_numbers, charges_mm, xyz_qm, xyz_mm):
+        """
+        Computes the static and induced EMLE energy components.
+
+        Parameters
+        ----------
+
+        atomic_numbers: torch.Tensor (N_QM_ATOMS,)
+            Atomic numbers of QM atoms.
+
+        charges_mm: torch.Tensor (max_mm_atoms,)
+            MM point charges in atomic units.
+
+        xyz_qm: torch.Tensor (N_QM_ATOMS, 3)
+            Positions of QM atoms in Angstrom.
+
+        xyz_mm: torch.Tensor (N_MM_ATOMS, 3)
+            Positions of MM atoms in Angstrom.
+
+        Returns
+        -------
+
+        result: torch.Tensor (3,)
+            The ANI2x and static and induced EMLE energy components in Hartree.
+        """
+        # Convert the atomic numbers to species IDs.
+        species_id = self._species_map[atomic_numbers]
+
+        # Reshape the IDs.
+        zid = species_id.unsqueeze(0)
+
+        # Reshape the coordinates,
+        xyz = xyz_qm.unsqueeze(0)
+
+        # Get the edge index and shifts for this configuration.
+        edge_index, shifts = self._get_neighbor_pairs(xyz_qm, None)
+
+        # Get the device.
+        device = xyz_qm.device
+
+        if self._atomic_numbers is None:
```

Comment:
Another way to do this, which would make the API look more like the other classes, is to set null attributes for the `self._node_attrs` and `self._atomic_numbers` in the constructor, then use `torch.eq` to see if the `atomic_numbers` that are passed to forward match the current `self._atomic_numbers`. If not, then the `_node_attrs` are recreated and `self._atomic_numbers` is set to `atomic_numbers`.

That said, I think your approach is fine and it probably makes sense to specialise an instance for a particular QM region via the constructor. The same thing is needed for `ANI2xEMLE` when using `NNPOps` anyway, since the optimisations assume a fixed QM region. (I don't check this in forward at present, i.e. the `atomic_numbers` passed in could differ.)

Reply:
I like your suggestion. If the QM region is fixed, then I think there's no need to recalculate everything at each iteration. I've revised this part of the code.